README.md
Jazz Hands [![Bitdeli Badge](https://d2weczhvl823v0.cloudfront.net/nixme/jazz_hands/trend.png)](https://bitdeli.com/free "Bitdeli Badge")
==========

Spending hours in the rails console? Spruce it up and show off those
hard-working hands!

**jazz_hands** is an opinionated set of console-related gems and a bit of glue:

* [**Pry**][pry] for a powerful shell alternative to IRB.
* [**Awesome Print**][awesome_print] for stylish pretty print.
* [**Hirb**][hirb] for tabular collection output.
* [**Pry Rails**][pry-rails] for additional commands (`show-routes`,
  `show-models`, `show-middleware`) in the Rails console.
* [**Pry Doc**][pry-doc] to browse Ruby source, including C, directly from the
  console.
* [**Pry Git**][pry-git] to teach the console about git. Diffs, blames, and
  commits on methods and classes, not just files.
* [**Pry Remote**][pry-remote] to connect remotely to a Pry console.
* [**Pry Debugger**][pry-debugger] to turn the console into a simple debugger.
* [**Pry Stack Explorer**][pry-stack_explorer] to navigate the call stack and
  frames.
* [**Coolline**][coolline] and [**Coderay**][coderay] for syntax highlighting as
  you type. _Optional. MRI 1.9.3/2.0.0 only_


## Usage

Ruby 1.9.2+, Rails 3 or 4 only. Add to your project Gemfile:

```ruby
group :development, :test do
  gem 'jazz_hands'
end
```

That's it. Run `rails console` as usual.

[Hirb][hirb] isn't enabled by default. To use, run `Hirb.enable` in the console.

Ruby compiled against a proper readline library, ideally GNU readline, is
recommended. Alternatively, [`gem install rb-readline`][rb-readline] for an
acceptable backup. Using ruby compiled against a `libedit` wrapper (primarily OS
X) will work but is not recommended.


## Options

Change the following options by creating an initializer in your Rails project
Example `config/initializers/jazz_hands.rb`:

```ruby
if defined?(JazzHands)
  JazzHands.colored_prompt = false
  JazzHands.enable_syntax_highlighting_as_you_type!
end
```

### `colored_prompt`

Color the console prompt? Defaults to `true` when the current ruby is compiled
against GNU readline or `rb-readline`, which don't have issues counting
characters in colored prompts. `false` for libedit.

Note: `Pry.color = false` trumps this setting and disables all console coloring.

### `prompt_separator`

Separator string between the application name and line input. Defaults to `»`
for GNU readline or libedit. Defaults to `>` for `rb-readline` which fails on
mixed encodings.

### Syntax highlighting

Syntax highlighting as you type via [Coolline][coolline] and [Coderay][coderay]
is disabled by default due to slightly buggy behavior. To enable, add
`JazzHands.enable_syntax_highlighting_as_you_type!` to the initializer. Only
works with MRI 1.9.3 or 2.0.0.


## Contributing

Patches and bug reports are welcome. Just send a [pull request][pullrequests] or
file an [issue][issues]. [Project changelog][changelog].


[pry]:                http://pry.github.com
[awesome_print]:      https://github.com/michaeldv/awesome_print
[hirb]:               https://github.com/cldwalker/hirb
[pry-rails]:          https://github.com/rweng/pry-rails
[pry-doc]:            https://github.com/pry/pry-doc
[pry-git]:            https://github.com/pry/pry-git
[pry-debugger]:       https://github.com/nixme/pry-debugger
[pry-remote]:         https://github.com/Mon-Ouie/pry-remote
[pry-stack_explorer]: https://github.com/pry/pry-stack_explorer
[coolline]:           https://github.com/Mon-Ouie/coolline
[coderay]:            https://github.com/rubychan/coderay
[rb-readline]:        https://github.com/luislavena/rb-readline
[pullrequests]:       https://github.com/nixme/jazz_hands/pulls
[issues]:             https://github.com/nixme/jazz_hands/issues
[changelog]:          https://github.com/nixme/jazz_hands/blob/master/CHANGELOG.md


CHANGELOG.md
## next (2023-06-23)

* Added support for the new module_parent_name method, required by
  newer Rails versions (>= 6.1)

## 1.4.0 (2020-04-27)

* Requires a more recent Pry version (at least 0.12.0) to work

## 1.3.0 (2020-01-09)

* Opened the dependency version requirements further

## 1.2.0 (2018-09-04)

* Opened the dependency version requirements

## 1.1.0 (2017-12-13)

* Removed [pry-remote][pry-remote]
* Added [pry-remote-reloaded][pry-remote-reloaded]

## 1.0.0 (2017-12-11)

* Removed [pry-stack_explorer][pry-stack_explorer]
* Removed [pry-git][pry-git]
* Removed [pry-debugger][pry-debugger]
* Added [pry-byebug][pry-byebug]
* Upgraded all dependencies
* Fixed the custom Pry.config.print proc
* Improved the Pry prompt

## 0.5.2 (2013-10-24)

* Upgrade [pry-rails][pry-rails] to the latest 0.3.2,
  [awesome_print][awesome_print] to 1.2 and loosen it's dependency, and removed
  [coderay][coderay] as an explicit dependency since [pry][pry] already includes
  it.

## 0.5.1 (2013-06-28)

* Upgrade [pry-rails][pry-rails] to the latest 0.3.1, [pry-doc][pry-doc] 0.4.6,
  [coolline][coolline] 0.4.2.


## 0.5.0 (2013-03-13)

* Rails 4, Ruby 2.0.0 compatibility.
* Upgrade [pry][pry] to the latest 0.9.12,
  [pry-stack_explorer][pry-stack_explorer] 0.4.9, [pry-remote][pry-remote]
  0.1.7, [pry-debugger][pry-debugger] 0.2.2, [hirb][hirb] 0.7.1,
  [coderay][coderay] 1.0.9.


## 0.4.0 (2012-12-27)

* Add [pry-rails][pry-rails] 0.2.2 for maintained console hooks and new
  `show-routes`, `show-models`, and `show-middleware` commands.
* Add `JazzHands.colored_prompt` and `JazzHands.prompt_separator` options.
  Detect readline library to turn off incompatible colored prompt or Unicode
  prompt separator where appropriate. Fixes #1 and #2.
* Upgrade [pry][pry] to the latest 0.9.10, [pry-doc][pry-doc] 0.4.4,
  [pry-stack_explorer][pry-stack_explorer] 0.4.7, [pry-remote][pry-remote]
  0.1.6, [pry-debugger][pry-debugger] 0.2.1, [hirb][hirb] 0.7.0,
  [Coolline][coolline] 0.4.0, [coderay][coderay] 1.0.8,
  [awesome_print][awesome_print] 1.1.0. Fixes #4.

## 0.3.1 (2012-06-11)

* Upgrade [pry-debugger][pry-debugger] to 0.2.0.

## 0.3.0 (2012-06-07)

* Replace [pry-nav][pry-nav] with [pry-debugger][pry-debugger] for improved
  performance and no segfaults. `JazzHands.enable_pry_nav` removed.
* Upgrade [pry][pry] to the latest 0.9.9, [pry-doc][pry-doc] 0.4.2,
  [pry-git][pry-git] 0.2.3, [pry-stack_explorer][pry-stack_explorer] 0.4.2,
  [pry-remote][pry-remote] 0.1.4, [coderay][coderay] 1.0.6.

## 0.2.0 (2012-03-19)

* Add [pry-stack_explorer][pry-stack_explorer] 0.4.1.
* Upgrade [pry-remote][pry-remote] to 0.1.1 and [pry-nav][pry-nav] to 0.2.0.
* Upgrade recommended gem minor versions: [pry][pry] 0.9.8.4, [pry-doc][pry-doc]
  0.4.1, [hirb][hirb] 0.6.2, and [coderay][coderay] 1.0.5.

## 0.1.2 (2012-01-23)

* Improved Rails 3.2 compatibility. Console methods like `app`, `new_session`,
  `reload!`, `helper`, and `controller` work as expected.

## 0.1.1 (2012-01-20)

* Rails 3.2 compatibility

## 0.1.0 (2012-01-04)

* For performance, enable [pry-nav][pry-nav] only on MRI 1.9.3 by default. To
  use on MRI 1.9.2, add `JazzHands.enable_pry_nav` to a Rails initializer.
* Due to buggy behavior, syntax highlighting as you type via
  [Coolline][coolline] and [Coderay][coderay] is disabled by default. Enable
  with `JazzHands.enable_syntax_highlighting_as_you_type` in a Rails
  initializer. MRI 1.9.3 only.
* Fix [Hirb][hirb] support.
* Upgrade [awesome_print][awesome_print] to 1.0.2.

## 0.0.6 (2011-12-03)

* Add line numbers to the prompt for easy reference when using `_in_` and
  `_out`.
* Upgrade [pry-nav][pry-nav] to 0.0.4.


## 0.0.5 (2011-12-01)

* Add [Hirb][hirb] support. Enable with `Hirb.enable` in the console.
* Upgrade [pry-nav][pry-nav] to 0.0.3.


## 0.0.4 (2011-11-30)

* Add explicit requires for pry plugin gems.


## 0.0.3 (2011-11-30)

* Add [pry-nav][pry-nav].
* Upgrade [awesome_print][awesome_print] to 1.0.1.


## 0.0.2 (2011-11-25)

* Add [pry-doc][pry-doc].


## 0.0.1 (2011-11-25)

* First release. Combine [pry][pry], [awesome_print][awesome_print],
  [coolline][coolline] + [coderay][coderay], [pry-remote][pry-remote],
  [pry-git][pry-git]. Bit of glue to replace IRB with pry in Rails console,
  pretty colors.

[pry]:                    http://pry.github.com
[awesome_print]:          https://github.com/michaeldv/awesome_print
[pry-doc]:                https://github.com/pry/pry-doc
[pry-git]:                https://github.com/pry/pry-git
[pry-nav]:                https://github.com/nixme/pry-nav
[pry-remote]:             https://github.com/Mon-Ouie/pry-remote
[coolline]:               https://github.com/Mon-Ouie/coolline
[coderay]:                https://github.com/rubychan/coderay
[hirb]:                   https://github.com/cldwalker/hirb
[pry-stack_explorer]:     https://github.com/pry/pry-stack_explorer
[pry-debugger]:           https://github.com/nixme/pry-debugger
[pry-rails]:              https://github.com/rweng/pry-rails
[pry-byebug]:             https://github.com/deivid-rodriguez/pry-byebug
[pry-remote-reloaded]:    https://github.com/Jack12816/pry-remote-reloaded


lib/funk_hands.rb
require 'funk_hands/version'
require 'funk_hands/railtie' if defined?(Rails)
require 'active_support'
require 'readline'

module FunkHands

  ### Options ###

  # Color the prompt?
  #
  # A different setting than Pry.color since some may like colored output, but a
  # plain prompt.
  #
  # Default: 'true' for GNU readline or rb-readline which correctly count line
  # widths with color codes when using \001 and \002 hints. 'false' for
  # libedit-based wrapper (standard on OS X unless ruby is explicitly compiled
  # otherwise).
  #
  mattr_accessor :colored_prompt
  self.colored_prompt = (Readline::VERSION !~ /EditLine/)

  # Separator between application name and input in the prompt.
  #
  # Default: right angle quote, or '>' when using rb-readline which doesn't
  # handle mixed encodings well.
  #
  mattr_accessor :prompt_separator
  self.prompt_separator = defined?(RbReadline) ? '>' : "\u00BB"


  class << self
    # Enable syntax highlighting as you type in the Rails console via coolline and
    # coderay (MRI 1.9.3+ only). Disabled by default as it's a bit buggy.
    #
    # Call from a Rails initializer:
    #
    #   FunkHands.enable_syntax_highlighting_as_you_type!
    #
    def enable_syntax_highlighting_as_you_type!
      raise 'Syntax highlighting only supported on 1.9.3+' unless RUBY_VERSION >= '1.9.3'

      # Use coolline with CodeRay for syntax highlighting as you type.
      # Only works on >= 1.9.3 because coolline depends on io/console.

      require 'coolline'
      require 'coderay'

      Pry.config.input = Coolline.new do |c|
        c.transform_proc = proc do
          CodeRay.scan(c.line, :ruby).term
        end

        c.completion_proc = proc do
          word = c.completed_word
          Object.constants.map(&:to_s).select { |w| w.start_with? word }
        end
      end
    end
    alias :enable_syntax_highlighting_as_you_type :enable_syntax_highlighting_as_you_type!
  end

  ### Internal methods ###

  mattr_accessor :_hirb_output
end


lib/jazz_hands.rb
require 'jazz_hands/version'
require 'jazz_hands/railtie' if defined?(Rails)
require 'active_support'
require 'readline'

module JazzHands

  ### Options ###

  # Color the prompt?
  #
  # A different setting than Pry.color since some may like colored output, but a
  # plain prompt.
  #
  # Default: 'true' for GNU readline or rb-readline which correctly count line
  # widths with color codes when using \001 and \002 hints. 'false' for
  # libedit-based wrapper (standard on OS X unless ruby is explicitly compiled
  # otherwise).
  #
  mattr_accessor :colored_prompt
  self.colored_prompt = (Readline::VERSION !~ /EditLine/)

  # Separator between application name and input in the prompt.
  #
  # Default: right angle quote, or '>' when using rb-readline which doesn't
  # handle mixed encodings well.
  #
  mattr_accessor :prompt_separator
  self.prompt_separator = defined?(RbReadline) ? '>' : "\u00BB"

  # Include the Rails.env in the prompt?
  #
  # Default: false
  #
  mattr_accessor :rails_env_in_prompt
  self.rails_env_in_prompt = false


  class << self
    # Enable syntax highlighting as you type in the Rails console via coolline and
    # coderay (MRI 1.9.3+ only). Disabled by default as it's a bit buggy.
    #
    # Call from a Rails initializer:
    #
    #   JazzHands.enable_syntax_highlighting_as_you_type!
    #
    def enable_syntax_highlighting_as_you_type!
      raise 'Syntax highlighting only supported on 1.9.3+' unless RUBY_VERSION >= '1.9.3'

      # Use coolline with CodeRay for syntax highlighting as you type.
      # Only works on >= 1.9.3 because coolline depends on io/console.

      require 'coolline'
      require 'coderay'

      Pry.config.input = Coolline.new do |c|
        c.transform_proc = proc do
          CodeRay.scan(c.line, :ruby).term
        end

        c.completion_proc = proc do
          word = c.completed_word
          Object.constants.map(&:to_s).select { |w| w.start_with? word }
        end
      end
    end
    alias :enable_syntax_highlighting_as_you_type :enable_syntax_highlighting_as_you_type!
  end

  ### Internal methods ###

  mattr_accessor :_hirb_output
end


lib/funk_hands/railtie.rb
require 'pry'
require 'pry-rails'
require 'pry-doc'
require 'pry-remote-reloaded'
require 'awesome_print'
require 'funk_hands/hirb_ext'
require 'pry-byebug'

module FunkHands
  class Railtie < Rails::Railtie
    initializer 'Funk_hands.initialize' do |app|
      silence_warnings do
        # We're managing the loading of plugins. So don't let pry autoload them.
        Pry.config.should_load_plugins = false

        # Use awesome_print for output, but keep pry's pager. If Hirb is
        # enabled, try printing with it first.
        # See: https://github.com/pry/pry/blob/v0.11.3/lib/pry.rb#L20
        Pry.config.print = ->(output, value, _pry_) do
          if FunkHands._hirb_output
            return output.puts(Hirb::View.view_or_page_output(value))
          end
          if Pry.pager
            return Pry::DEFAULT_PRINT.call(output, value, _pry_)
          end
          output.puts("=> #{value.ai(indent: 2)}")
        end

        # Friendlier prompt - line number, app name, nesting levels look like
        # directory paths.
        #
        # Heavy use of lazy lambdas so configuration (like Pry.color) can be
        # changed later or even during console usage.
        #
        # Custom color helpers using hints \001 and \002 so that good readline
        # libraries (GNU, rb-readline) correctly ignore color codes when
        # calculating line length.

        color = -> { Pry.color && FunkHands.colored_prompt }
        red  = ->(text) { color[] ? "\001\e[0;31m\002#{text}\001\e[0m\002" : text.to_s }
        blue = ->(text) { color[] ? "\001\e[0;34m\002#{text}\001\e[0m\002" : text.to_s }
        bold = ->(text) { color[] ? "\001\e[1m\002#{text}\001\e[0m\002"    : text.to_s }

        name = app.class.module_parent_name \
          if app.class.respond_to?(:module_parent_name)
        name ||= app.class.parent_name
        name = name.underscore

        separator = -> { red.(FunkHands.prompt_separator) }
        colored_name = -> { blue.(name) }

        line = ->(pry) { "[#{bold.(pry.input_ring.size)}] " }
        target_string = ->(object, level) do
          level = 0 if level < 0
          unless (string = Pry.view_clip(object)) == 'main'
            "(#{'../' * level}#{string})"
          else
            ''
          end
        end

        Pry.config.prompt = [
          ->(object, level, pry) do      # Main prompt
            "#{line.(pry)}#{colored_name.()}#{target_string.(object, level)} #{separator.()} "
          end,
          ->(object, level, pry) do      # Wait prompt in multiline input
            spaces = ' ' * (
              "[#{pry.input_ring.size}] ".size +  # Uncolored `line.(pry)`
              name.size +
              target_string.(object, level).size
            )
            "#{spaces} #{separator.()} "
          end
        ]
      end
    end
  end
end
